sys/mips/ebase.S

```diff
@@ -345,15 +347,14 @@ SNESTED(user_exc_enter, MCONTEXT_SIZE, ra)
         .cfi_rel_offset t1, CTX_PC
         PCPU_TO_FRAME(BADVADDR)
 
-        # If FPU is enabled save FPU registers.
-        mfc0    t0, C0_STATUS
-        ext     t1, t0, SR_CU1_SHIFT, 1
-        beqz    t1, skip_fpu_save
-        nop
-
-        SAVE_FPU_CTX()
+        # Invalidate current FPU context.
+        LOAD_PCPU(t0)
+        lw      t0, PCPU_CURTHREAD(t0)
+        lw      t1, TD_PFLAGS(t0)
+        nor     t2, zero, TDP_FPCTXSAVED
+        and     t1, t1, t2
+        sw      t1, TD_PFLAGS(t0)
```

Comment:
Couldn't we clear the `TDP_FPCTXSAVED` flag in `user_exc_leave()`, in the same place where we restore the FPU context? That way, we do this only for threads that have the flag set.

Reply:
done